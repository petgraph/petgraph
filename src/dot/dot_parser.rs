use crate::data::Create;
use alloc::boxed::Box;
use core::convert::TryFrom;
use core::error::Error;
use core::fmt::{Display, Formatter};
use dot_parser::ast::AList;
use dot_parser::ast::Graph as DotGraph;
use dot_parser::ast::PestError as ParsingError;
use dot_parser::canonical::Graph as CGraph;
use dot_parser::canonical::Node;

pub type DotNodeWeight<'a> = Node<(&'a str, &'a str)>;
pub type DotAttrList<'a> = AList<(&'a str, &'a str)>;

#[derive(Debug)]
pub struct DotParsingError {
    error: Box<ParsingError>,
}

impl Display for DotParsingError {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), core::fmt::Error> {
        write!(f, "{}", self.error)
    }
}

impl From<ParsingError> for DotParsingError {
    fn from(error: ParsingError) -> Self {
        Self {
            error: Box::new(error),
        }
    }
}

impl Error for DotParsingError {}

/// This trait extends [Create] with a method to parse a graph from a dot string.
pub trait ParseFromDot<'a>:
    Create<EdgeWeight = DotAttrList<'a>, NodeWeight = DotNodeWeight<'a>>
{
    /// Convert a DOT/Graphviz graph (represented as an [DotGraph]) into a petgraph's graph.
    fn from_dot_graph(dot_graph: DotGraph<(&'a str, &'a str)>) -> Self {
        let dot_graph: CGraph<(&'a str, &'a str)> = dot_graph.into();
        let node_number = dot_graph.nodes.set.len();
        let edge_number = dot_graph.edges.set.len();
        let mut graph = Self::with_capacity(node_number, edge_number);
        let mut node_indices = std::collections::HashMap::new();
        for node in dot_graph.nodes.set {
            let ni = graph.add_node(node.1);
            node_indices.insert(node.0, ni);
        }
        for edge in dot_graph.edges.set {
            let from_ni = node_indices.get(&edge.from).unwrap();
            let to_ni = node_indices.get(&edge.to).unwrap();
            graph.add_edge(*from_ni, *to_ni, edge.attr);
        }
        graph
    }

    /// Attempt to parse a DOT/Graphviz string into a graph. Fail if the string is not a
    /// well-formed DOT/Graphviz string.
    fn try_from(s: &'a str) -> Result<Self, DotParsingError> {
        let ast = DotGraph::try_from(s)?;
        let petgraph = Self::from_dot_graph(ast);
        Ok(petgraph)
    }
}

#[macro_export]
/// Statically imports a Graph from a valid DOT/Graphviz [&str].
macro_rules! graph_from_str {
    ($s:tt) => {
        $crate::dot::dot_parser::ParseFromDot::from_dot_graph(dot_parser_macros::from_dot_string!(
            $s
        ))
    };
}

#[macro_export]
/// Statically imports a Graph from a DOT/Graphviz file. The macro expects the file path as argument.
///
/// Notice that, since the graph is imported *statically*, the file must exist at compile time, but
/// can be removed at runtime.
macro_rules! graph_from_file {
    ($s:tt) => {
        $crate::dot::dot_parser::ParseFromDot::from_dot_graph(dot_parser_macros::from_dot_file!($s))
    };
}

pub use graph_from_file;
pub use graph_from_str;

impl<'a> ParseFromDot<'a> for crate::graph::Graph<DotNodeWeight<'a>, DotAttrList<'a>> {}
#[cfg(feature = "stable_graph")]
impl<'a> ParseFromDot<'a> for crate::stable_graph::StableGraph<DotNodeWeight<'a>, DotAttrList<'a>> {}

#[cfg(test)]
mod test {
    #[test]
    fn test_dot_parsing_str() {
        let _: crate::graph::Graph<_, _> = graph_from_str!("digraph { A -> B }");
        #[cfg(feature = "stable_graph")]
        let _: crate::stable_graph::StableGraph<_, _> = graph_from_str!("digraph { A -> B }");
    }

    #[test]
    fn test_ill_formed_str() {
        let g_res: Result<crate::graph::Graph<_, _>, crate::dot::dot_parser::DotParsingError> =
            crate::dot::dot_parser::ParseFromDot::try_from(":zcdza");
        assert!(g_res.is_err())
    }

    #[test]
    fn test_dot_parsing_file() {
        let _: crate::graph::Graph<_, _> = graph_from_file!("graph-example.dot");
        #[cfg(feature = "stable_graph")]
        let _: crate::stable_graph::StableGraph<_, _> = graph_from_file!("graph-example.dot");
    }

    #[test]
    fn test_dot_parsing_isomorph() {
        use crate::algo::is_isomorphic;

        let g1: crate::graph::Graph<_, _> = graph_from_str!("digraph { A -> { B C }}");
        let g2: crate::graph::Graph<_, _> = graph_from_str!("digraph { D -> E; D -> F}");
        assert!(is_isomorphic(&g1, &g2));
    }

    #[test]
    fn test_dot_parsing_not_isomorph() {
        use crate::algo::is_isomorphic;

        let g1: crate::graph::Graph<_, _> = graph_from_str!("digraph { A -> { B C D }}");
        let g2: crate::graph::Graph<_, _> = graph_from_str!("digraph { A -> B; A -> C}");
        assert!(!is_isomorphic(&g1, &g2));
    }

    #[test]
    fn test_dot_parsing_subgraph_isomorph() {
        use crate::algo::is_isomorphic_subgraph;

        let g1: crate::graph::Graph<_, _> = graph_from_str!("digraph { A -> B; A -> C}");
        let g2: crate::graph::Graph<_, _> = graph_from_str!("digraph { A -> { B C D }}");
        assert!(is_isomorphic_subgraph(&g1, &g2));
    }

    #[test]
    fn test_dot_parsing_cyclic() {
        use crate::algo::is_cyclic_directed;
        let g: crate::graph::Graph<_, _> =
            graph_from_str!("digraph { A -> { B C } -> {D E F}; F -> A}");
        assert!(is_cyclic_directed(&g));
    }

    #[test]
    fn test_dot_parsing_not_cyclic() {
        use crate::algo::is_cyclic_directed;
        let g: crate::graph::Graph<_, _> = graph_from_str!("digraph { A -> { B C } -> {D E F}}");
        assert!(!is_cyclic_directed(&g));
    }
}
