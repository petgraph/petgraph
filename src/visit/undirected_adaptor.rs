use crate::visit::{
    Data, GraphBase, GraphProp, GraphRef, IntoEdgeReferences, IntoEdges, IntoEdgesDirected,
    IntoNeighbors, IntoNeighborsDirected, IntoNodeIdentifiers, IntoNodeReferences,
    NodeCompactIndexable, NodeCount, NodeIndexable, Visitable,
};
use crate::Direction;

/// An edge direction removing graph adaptor.
#[derive(Copy, Clone, Debug)]
pub struct UndirectedAdaptor<G>(G);

impl<G: GraphRef> GraphRef for UndirectedAdaptor<G> {}

impl<G> IntoNeighbors for UndirectedAdaptor<G>
where
    G: IntoNeighborsDirected,
{
    type Neighbors = std::iter::Chain<G::NeighborsDirected, G::NeighborsDirected>;
    fn neighbors(self, n: G::NodeId) -> Self::Neighbors {
        self.0
            .neighbors_directed(n, Direction::Incoming)
            .chain(self.0.neighbors_directed(n, Direction::Outgoing))
    }
}

impl<G> IntoEdges for UndirectedAdaptor<G>
where
    G: IntoEdgesDirected,
{
    type Edges = std::iter::Chain<G::EdgesDirected, G::EdgesDirected>;
    fn edges(self, a: Self::NodeId) -> Self::Edges {
        self.0
            .edges_directed(a, Direction::Incoming)
            .chain(self.0.edges_directed(a, Direction::Outgoing))
    }
}

impl<G> GraphProp for UndirectedAdaptor<G>
where
    G: GraphBase,
{
    type EdgeType = crate::Undirected;

    fn is_directed(&self) -> bool {
        false
    }
}

macro_rules! access0 {
    ($e:expr) => {
        $e.0
    };
}

GraphBase! {delegate_impl [[G], G, UndirectedAdaptor<G>, access0]}
Data! {delegate_impl [[G], G, UndirectedAdaptor<G>, access0]}
IntoEdgeReferences! {delegate_impl [[G], G, UndirectedAdaptor<G>, access0]}
Visitable! {delegate_impl [[G], G, UndirectedAdaptor<G>, access0]}
NodeIndexable! {delegate_impl [[G], G, UndirectedAdaptor<G>, access0]}
NodeCompactIndexable! {delegate_impl [[G], G, UndirectedAdaptor<G>, access0]}
IntoNodeIdentifiers! {delegate_impl [[G], G, UndirectedAdaptor<G>, access0]}
IntoNodeReferences! {delegate_impl [[G], G, UndirectedAdaptor<G>, access0]}
NodeCount! {delegate_impl [[G], G, UndirectedAdaptor<G>, access0]}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::graph::{DiGraph, Graph};
    use crate::visit::Dfs;

    static LINEAR_EDGES: [(u32, u32); 5] = [(0, 1), (1, 2), (2, 3), (3, 4), (4, 5)];

    #[test]
    pub fn test_is_reachable() {
        // create a linear digraph, choose a node in the centre and check all nodes are visited
        // by a dfs

        let graph = DiGraph::<(), ()>::from_edges(&LINEAR_EDGES);

        let mut nodes = graph.node_identifiers().collect::<Vec<_>>();
        nodes.sort();

        let graph = UndirectedAdaptor(&graph);

        use crate::visit::Walker;
        let mut visited_nodes: Vec<_> = Dfs::new(&graph, nodes[2]).iter(&graph).collect();
        visited_nodes.sort();
        assert_eq!(visited_nodes, nodes);
    }

    #[test]
    pub fn test_neighbors_count() {
        {
            let graph = Graph::<(), ()>::from_edges(&LINEAR_EDGES);
            let graph = UndirectedAdaptor(&graph);

            let mut nodes = graph.node_identifiers().collect::<Vec<_>>();
            nodes.sort();
            assert_eq!(graph.neighbors(nodes[1]).count(), 2);
        }

        {
            let graph = Graph::<(), ()>::from_edges(&LINEAR_EDGES);
            let graph = UndirectedAdaptor(&graph);

            let mut nodes = graph.node_identifiers().collect::<Vec<_>>();
            nodes.sort();
            assert_eq!(graph.neighbors(nodes[1]).count(), 2);
        }
    }
}
