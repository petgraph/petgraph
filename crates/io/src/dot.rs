//! Simple graphviz dot file format output.
// TODO: potentially move this into petgraph-io or petgraph-dot (io for different formats)

use std::{
    borrow::Cow,
    fmt::{self, Debug, Display, Formatter},
};

pub use dot::RenderOption;
use dot::{Arrow, Edges, Id, Kind, LabelText, Nodes, Style};
use petgraph_core::visit::{
    EdgeRef, GraphProp, IntoEdgeReferences, IntoNodeIdentifiers, IntoNodeReferences, NodeIndexable,
    NodeRef,
};

pub struct NodeAttributes {
    pub label: LabelText<'static>,
    pub style: Style,

    pub color: Option<LabelText<'static>>,

    pub shape: Option<LabelText<'static>>,
}

impl NodeAttributes {
    #[must_use]
    pub fn new(label: impl Into<String>) -> Self {
        Self {
            label: LabelText::LabelStr(Cow::Owned(label.into())),
            style: Style::None,

            color: None,

            shape: None,
        }
    }

    #[must_use]
    pub const fn with_style(mut self, style: Style) -> Self {
        self.style = style;
        self
    }

    #[must_use]
    pub fn with_color(mut self, color: impl Into<LabelText<'static>>) -> Self {
        self.color = Some(color.into());
        self
    }

    #[must_use]
    pub fn with_shape(mut self, shape: impl Into<LabelText<'static>>) -> Self {
        self.shape = Some(shape.into());
        self
    }
}

pub struct EdgeAttributes {
    pub label: LabelText<'static>,
    pub style: Style,

    pub color: Option<LabelText<'static>>,

    pub start_arrow: Arrow,
    pub end_arrow: Arrow,
}

impl EdgeAttributes {
    #[must_use]
    pub fn new(label: impl Into<String>) -> Self {
        Self {
            label: LabelText::LabelStr(Cow::Owned(label.into())),
            style: Style::None,

            color: None,

            start_arrow: Arrow::none(),
            end_arrow: Arrow::normal(),
        }
    }

    #[must_use]
    pub const fn with_style(mut self, style: Style) -> Self {
        self.style = style;
        self
    }

    #[must_use]
    pub fn with_color(mut self, color: impl Into<LabelText<'static>>) -> Self {
        self.color = Some(color.into());
        self
    }

    #[allow(clippy::missing_const_for_fn)] // Reason: false positive
    #[must_use]
    pub fn with_start_arrow(mut self, arrow: Arrow) -> Self {
        self.start_arrow = arrow;
        self
    }

    #[allow(clippy::missing_const_for_fn)] // Reason: false positive
    #[must_use]
    pub fn with_end_arrow(mut self, arrow: Arrow) -> Self {
        self.end_arrow = arrow;
        self
    }
}

/// `Dot` implements output to graphviz .dot format for a graph.
///
/// Formatting and options are rather simple, this is mostly intended
/// for debugging. Exact output may change.
///
/// # Examples
///
/// ```
/// use dot::RenderOption;
/// use petgraph::{graph::Graph, io::dot::Dot};
///
/// let mut graph = Graph::<_, ()>::new();
/// graph.add_node("A");
/// graph.add_node("B");
/// graph.add_node("C");
/// graph.add_node("D");
/// graph.extend_with_edges(&[(0, 1), (0, 2), (0, 3), (1, 2), (1, 3), (2, 3)]);
///
/// println!(
///     "{:?}",
///     Dot::with_config(&graph, &[RenderOption::NoEdgeLabels])
/// );
///
/// // In this case the output looks like this:
/// //
/// // digraph {
/// //     0 [label="\"A\""]
/// //     1 [label="\"B\""]
/// //     2 [label="\"C\""]
/// //     3 [label="\"D\""]
/// //     0 -> 1
/// //     0 -> 2
/// //     0 -> 3
/// //     1 -> 2
/// //     1 -> 3
/// //     2 -> 3
/// // }
///
/// // If you need multiple config options, just list them all in the slice.
/// ```
// TODO: revisit this API, it's not very ergonomic
pub struct Dot<'a, G>
where
    G: IntoEdgeReferences + IntoNodeReferences,
{
    graph: G,

    get_node_attributes: &'a dyn Fn(G, G::NodeRef) -> NodeAttributes,
    get_edge_attributes: &'a dyn Fn(G, G::EdgeRef) -> EdgeAttributes,

    options: &'a [RenderOption],
    // in future we might want to cache node and edge attributes, not implemented yet because we
    // would need to tighten the bounds, which we will change anyway
}

impl<'a, G> Dot<'a, G>
where
    G: IntoEdgeReferences + IntoNodeReferences,
{
    fn call_node_attributes(&self, node: G::NodeId) -> NodeAttributes {
        let node = self
            .graph
            .node_references()
            .find(|n| n.id() == node)
            .expect("node not found");

        (self.get_node_attributes)(self.graph, node)
    }

    fn call_edge_attributes(&self, edge: G::EdgeRef) -> EdgeAttributes {
        (self.get_edge_attributes)(self.graph, edge)
    }
}

impl<'a, G> Dot<'a, G>
where
    G: IntoNodeIdentifiers + IntoNodeReferences + IntoEdgeReferences,
    // TODO: remove this bound once we have reworked how indices work
    G::NodeId: Display,
{
    /// Create a `Dot` formatting wrapper with default configuration.
    #[inline]
    pub fn new(graph: G) -> Self {
        Self::with_config(graph, &[])
    }

    /// Create a `Dot` formatting wrapper with custom configuration.
    #[inline]
    pub fn with_config(graph: G, options: &'a [RenderOption]) -> Self {
        Self::with_attr_getters(
            graph, //
            options,
            &|_, _| EdgeAttributes::new(""),
            &|_, node| NodeAttributes::new(node.id().to_string()),
        )
    }

    #[inline]
    pub fn with_attr_getters(
        graph: G,
        options: &'a [RenderOption],

        get_edge_attributes: &'a dyn Fn(G, G::EdgeRef) -> EdgeAttributes,
        get_node_attributes: &'a dyn Fn(G, G::NodeRef) -> NodeAttributes,
    ) -> Self {
        Self {
            graph,

            get_node_attributes,
            get_edge_attributes,

            options,
        }
    }

    /// Set the function used to get the attributes for an edge.
    #[must_use]
    pub fn with_edge_attributes(
        mut self,
        get_edge_attributes: &'a dyn Fn(G, G::EdgeRef) -> EdgeAttributes,
    ) -> Self {
        self.get_edge_attributes = get_edge_attributes;
        self
    }

    /// Set the function used to get the attributes for a node.
    #[must_use]
    pub fn with_node_attributes(
        mut self,
        get_node_attributes: &'a dyn Fn(G, G::NodeRef) -> NodeAttributes,
    ) -> Self {
        self.get_node_attributes = get_node_attributes;
        self
    }

    /// Write the graph to the given writer in the dot format.
    ///
    /// This delegates most of the heavy lifting to the [`dot`] crate.
    ///
    /// # Errors
    ///
    /// This function will return an error if the underlying [`dot`] crate fails to write to the
    /// given writer.
    ///
    /// [`dot`]: https://docs.rs/dot
    pub fn write<W>(self, write: &mut W) -> std::io::Result<()>
    where
        W: std::io::Write,
    {
        dot::render_opts(&self, write, self.options)
    }
}

impl<'a, G> dot::Labeller<'a, G::NodeId, G::EdgeRef> for Dot<'a, G>
where
    G: IntoNodeIdentifiers + IntoNodeReferences + IntoEdgeReferences,
    G::NodeId: Display,
{
    fn graph_id(&'a self) -> Id<'a> {
        // TODO: make configurable
        Id::new("petgraph").expect("infallible")
    }

    fn node_id(&'a self, n: &G::NodeId) -> Id<'a> {
        Id::new(format!("N{n}")).expect("infallible")
    }

    fn node_label(&'a self, node: &G::NodeId) -> LabelText<'a> {
        self.call_node_attributes(*node).label
    }

    fn node_color(&'a self, node: &G::NodeId) -> Option<LabelText<'a>> {
        self.call_node_attributes(*node).color
    }

    fn node_style(&'a self, node: &G::NodeId) -> Style {
        self.call_node_attributes(*node).style
    }

    fn node_shape(&'a self, node: &G::NodeId) -> Option<LabelText<'a>> {
        self.call_node_attributes(*node).shape
    }

    fn edge_color(&'a self, edge: &G::EdgeRef) -> Option<LabelText<'a>> {
        self.call_edge_attributes(*edge).color
    }

    fn edge_end_arrow(&'a self, edge: &G::EdgeRef) -> Arrow {
        self.call_edge_attributes(*edge).end_arrow
    }

    fn edge_label(&'a self, edge: &G::EdgeRef) -> LabelText<'a> {
        self.call_edge_attributes(*edge).label
    }

    fn edge_start_arrow(&'a self, edge: &G::EdgeRef) -> Arrow {
        self.call_edge_attributes(*edge).start_arrow
    }

    fn edge_style(&'a self, edge: &G::EdgeRef) -> Style {
        self.call_edge_attributes(*edge).style
    }

    // TODO: we have no way to know if the graph is directed or not
    // fn kind(&self) -> Kind {
    //     if G::is_directed() {
    //         Kind::Digraph
    //     } else {
    //         Kind::Graph
    //     }
    // }
}

impl<'a, G> dot::GraphWalk<'a, G::NodeId, G::EdgeRef> for Dot<'a, G>
where
    G: IntoNodeIdentifiers + IntoNodeReferences + IntoEdgeReferences,
{
    fn nodes(&'a self) -> Nodes<'a, G::NodeId> {
        Nodes::Owned(self.graph.node_identifiers().collect())
    }

    fn edges(&'a self) -> Edges<'a, G::EdgeRef> {
        Edges::Owned(self.graph.edge_references().collect())
    }

    fn source(&'a self, edge: &G::EdgeRef) -> G::NodeId {
        edge.source()
    }

    fn target(&'a self, edge: &G::EdgeRef) -> G::NodeId {
        edge.target()
    }
}

impl<'a, G> Display for Dot<'a, G>
where
    G: IntoNodeReferences + IntoNodeReferences + IntoEdgeReferences,
    G::NodeId: Display,
{
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        Debug::fmt(self, f)
    }
}

impl<'a, G> Debug for Dot<'a, G>
where
    G: IntoNodeReferences + IntoNodeReferences + IntoEdgeReferences,
    G::NodeId: Display,
{
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        let mut buffer = Vec::new();

        dot::render_opts(self, &mut buffer, self.options).map_err(|_| fmt::Error)?;

        let mut s = String::from_utf8(buffer).map_err(|_| fmt::Error)?;
        f.write_str(&s)
    }
}
