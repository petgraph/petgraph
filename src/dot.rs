//! Simple graphviz dot file format output.
// TODO: potentially move this into petgraph-io or petgraph-dot (io for different formats)

use std::fmt::{self, Debug, Display, Formatter, Write};

use dot::{Edges, Id, LabelText, Nodes, RenderOption};

use crate::visit::{
    EdgeRef, GraphProp, IntoEdgeReferences, IntoNodeReferences, NodeIndexable, NodeRef,
};

/// `Dot` implements output to graphviz .dot format for a graph.
///
/// Formatting and options are rather simple, this is mostly intended
/// for debugging. Exact output may change.
///
/// # Examples
///
/// ```
/// use petgraph::{
///     dot::{Config, Dot},
///     Graph,
/// };
///
/// let mut graph = Graph::<_, ()>::new();
/// graph.add_node("A");
/// graph.add_node("B");
/// graph.add_node("C");
/// graph.add_node("D");
/// graph.extend_with_edges(&[(0, 1), (0, 2), (0, 3), (1, 2), (1, 3), (2, 3)]);
///
/// println!("{:?}", Dot::with_config(&graph, &[Config::EdgeNoLabel]));
///
/// // In this case the output looks like this:
/// //
/// // digraph {
/// //     0 [label="\"A\""]
/// //     1 [label="\"B\""]
/// //     2 [label="\"C\""]
/// //     3 [label="\"D\""]
/// //     0 -> 1
/// //     0 -> 2
/// //     0 -> 3
/// //     1 -> 2
/// //     1 -> 3
/// //     2 -> 3
/// // }
///
/// // If you need multiple config options, just list them all in the slice.
/// ```
pub struct Dot<'a, G>
where
    G: IntoEdgeReferences + IntoNodeReferences,
{
    graph: G,
    get_edge_attributes: &'a dyn Fn(G, G::EdgeRef) -> String,
    get_node_attributes: &'a dyn Fn(G, G::NodeRef) -> String,
    options: &'a [RenderOption],
}

static TYPE: [&str; 2] = ["graph", "digraph"];
static EDGE: [&str; 2] = ["--", "->"];
static INDENT: &str = "    ";

impl<'a, G> Dot<'a, G>
where
    G: IntoNodeReferences + IntoEdgeReferences,
{
    /// Create a `Dot` formatting wrapper with default configuration.
    #[inline]
    pub fn new(graph: G) -> Self {
        Self::with_config(graph, &[])
    }

    /// Create a `Dot` formatting wrapper with custom configuration.
    #[inline]
    pub fn with_config(graph: G, options: &'a [RenderOption]) -> Self {
        Self::with_attr_getters(
            graph, //
            &options,
            &|_, _| String::new(),
            &|_, _| String::new(),
        )
    }

    #[inline]
    pub fn with_attr_getters(
        graph: G,
        options: &'a [RenderOption],

        get_edge_attributes: &'a dyn Fn(G, G::EdgeRef) -> String,
        get_node_attributes: &'a dyn Fn(G, G::NodeRef) -> String,
    ) -> Self {
        Self {
            graph,
            options,

            get_edge_attributes,
            get_node_attributes,
        }
    }
}

impl<'a, G> dot::Labeller<'a, G::NodeId, G::EdgeRef> for Dot<'a, G>
where
    G: IntoNodeReferences + IntoEdgeReferences,
{
    fn graph_id(&'a self) -> Id<'a> {
        // TODO: make configurable
        Id::new("petgraph").expect("infallible")
    }

    fn node_id(&'a self, n: &G::NodeId) -> Id<'a> {
        Id::new(format!("N{}", n.as_usize())).expect("infallible")
    }

    fn node_label(&'a self, n: &G::NodeId) -> LabelText<'a> {
        // TODO: this needs improvement, as it is incomplete!
        let label = (self.get_node_attributes)(self.graph, self.graph.node_reference(*n));

        LabelText::LabelStr(label.into())
    }

    fn edge_label(&'a self, e: &G::EdgeRef) -> LabelText<'a> {
        let label = (self.get_edge_attributes)(self.graph, self.graph.edge_reference(e.id()));

        LabelText::LabelStr(label.into())
    }
}

impl<'a, G> dot::GraphWalk<'a, G::NodeId, G::EdgeRef> for Dot<'a, G>
where
    G: IntoNodeReferences + IntoEdgeReferences,
{
    fn nodes(&'a self) -> Nodes<'a, G::NodeId> {
        Nodes::Owned(self.graph.node_identifiers().collect())
    }

    fn edges(&'a self) -> Edges<'a, G::EdgeRef> {
        Edges::Owned(self.graph.edge_references().collect())
    }

    fn source(&'a self, edge: &G::EdgeRef) -> G::NodeId {
        edge.source()
    }

    fn target(&'a self, edge: &G::EdgeRef) -> G::NodeId {
        edge.target()
    }
}

impl<'a, G> Display for Dot<'a, G>
where
    G: IntoNodeReferences + IntoEdgeReferences,
{
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        dot::render_opts(self, f, self.options).map_err(|_| fmt::Error)
    }
}

impl<'a, G> Debug for Dot<'a, G>
where
    G: IntoNodeReferences + IntoEdgeReferences,
{
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        dot::render_opts(self, f, self.options).map_err(|_| fmt::Error)
    }
}

#[cfg(test)]
mod test {
    use std::fmt::Write;

    use dot::RenderOption;

    use super::{Config, Dot, Escaper};
    use crate::{prelude::Graph, visit::NodeRef};

    fn simple_graph() -> Graph<&'static str, &'static str> {
        let mut graph = Graph::<&str, &str>::new();
        let a = graph.add_node("A");
        let b = graph.add_node("B");
        graph.add_edge(a, b, "edge_label");
        graph
    }

    #[test]
    fn test_nodeindexlable_option() {
        let graph = simple_graph();
        let dot = format!("{:?}", Dot::with_config(&graph, &[]));
        assert_eq!(
            dot,
            "digraph {\n    0 [ label = \"0\" ]\n    1 [ label = \"1\" ]\n    0 -> 1 [ label = \
             \"\\\"edge_label\\\"\" ]\n}\n"
        );
    }

    #[test]
    fn test_edgeindexlable_option() {
        let graph = simple_graph();
        let dot = format!("{:?}", Dot::with_config(&graph, &[]));
        assert_eq!(
            dot,
            "digraph {\n    0 [ label = \"\\\"A\\\"\" ]\n    1 [ label = \"\\\"B\\\"\" ]\n    0 \
             -> 1 [ label = \"0\" ]\n}\n"
        );
    }

    #[test]
    fn test_edgenolable_option() {
        let graph = simple_graph();
        let dot = format!(
            "{:?}",
            Dot::with_config(&graph, &[RenderOption::NoEdgeLabels])
        );
        assert_eq!(
            dot,
            "digraph {\n    0 [ label = \"\\\"A\\\"\" ]\n    1 [ label = \"\\\"B\\\"\" ]\n    0 \
             -> 1 [ ]\n}\n"
        );
    }

    #[test]
    fn test_nodenolable_option() {
        let graph = simple_graph();
        let dot = format!(
            "{:?}",
            Dot::with_config(&graph, &[RenderOption::NoNodeLabels])
        );
        assert_eq!(
            dot,
            "digraph {\n    0 [ ]\n    1 [ ]\n    0 -> 1 [ label = \"\\\"edge_label\\\"\" ]\n}\n"
        );
    }

    #[test]
    fn test_with_attr_getters() {
        let graph = simple_graph();
        let dot = format!(
            "{:?}",
            Dot::with_attr_getters(
                &graph,
                &[RenderOption::NoNodeLabels, RenderOption::NoEdgeLabels],
                &|_, er| format!("label = \"{}\"", er.weight().to_uppercase()),
                &|_, nr| format!("label = \"{}\"", nr.weight().to_lowercase()),
            ),
        );
        assert_eq!(
            dot,
            "digraph {\n    0 [ label = \"a\"]\n    1 [ label = \"b\"]\n    0 -> 1 [ label = \
             \"EDGE_LABEL\"]\n}\n"
        );
    }
}
