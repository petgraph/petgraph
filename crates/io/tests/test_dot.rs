#![cfg(not(miri))]
#![cfg(feature = "dot")]

use dot::RenderOption;
use insta::assert_debug_snapshot;
use petgraph::{adjacency_matrix::AdjacencyList, graph::stable::StableGraph, Graph};
use petgraph_core::visit::EdgeRef;
use petgraph_io::dot::{Dot, EdgeAttributes, NodeAttributes};

fn simple_graph() -> Graph<&'static str, &'static str> {
    let mut graph = Graph::<&str, &str>::new();
    let a = graph.add_node("A");
    let b = graph.add_node("B");
    graph.add_edge(a, b, "edge_label");
    graph
}

#[test]
fn node_index_label() {
    let graph = simple_graph();

    let dot = Dot::with_config(&graph, &[])
        .with_node_attributes(&|_, (index, _)| NodeAttributes::new(index.to_string()));

    assert_debug_snapshot!(dot);
}

#[test]
fn edge_index_label() {
    let graph = simple_graph();

    let dot = Dot::with_config(&graph, &[])
        .with_edge_attributes(&|_, edge| EdgeAttributes::new(edge.id().index().to_string()));

    assert_debug_snapshot!(dot);
}

#[test]
fn edge_no_label() {
    let graph = simple_graph();
    let dot = Dot::with_config(&graph, &[RenderOption::NoEdgeLabels]);

    assert_debug_snapshot!(dot);
}

#[test]
fn node_no_label() {
    let graph = simple_graph();
    let dot = Dot::with_config(&graph, &[RenderOption::NoNodeLabels]);

    assert_debug_snapshot!(dot);
}

#[test]
fn label_map_to_weight() {
    let graph = simple_graph();
    let dot = Dot::with_config(&graph, &[])
        .with_node_attributes(&|_, (_, weight)| NodeAttributes::new(weight.to_uppercase()))
        .with_edge_attributes(&|_, edge| EdgeAttributes::new(edge.weight().to_uppercase()));

    assert_debug_snapshot!(dot);
}

#[test]
fn adjacency_list() {
    let mut graph = AdjacencyList::new();

    let a = graph.add_node();
    let b = graph.add_node();

    graph.add_edge(a, a, "A -> A");
    graph.add_edge(a, b, "A -> B");
    graph.add_edge(b, a, "B -> A");

    assert_debug_snapshot!(Dot::new(&graph));
}

#[test]
fn graph() {
    let mut graph = Graph::new();

    let a = graph.add_node("A");
    let b = graph.add_node("B");

    graph.add_edge(a, a, "A -> A");
    graph.add_edge(a, b, "A -> B");
    graph.add_edge(b, a, "B -> A");

    assert_debug_snapshot!(Dot::new(&graph));
}

#[test]
fn debug_formatting() {
    // test alternate formatting
    #[derive(Debug)]
    struct Record {
        a: i32,
        b: &'static str,
    }
    let mut graph = Graph::new();
    let a = graph.add_node(Record { a: 1, b: r"abc\" });
    graph.add_edge(a, a, (1, 2));

    assert_debug_snapshot!(Dot::new(&graph));
}

#[test]
fn stable_graph() {
    let mut graph = StableGraph::new();

    let a = graph.add_node("A");
    let b = graph.add_node("B");

    graph.add_edge(a, a, "A -> A");
    graph.add_edge(a, b, "A -> B");

    assert_debug_snapshot!(Dot::new(&graph));
}
