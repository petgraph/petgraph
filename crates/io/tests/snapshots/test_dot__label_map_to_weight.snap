---
source: crates/io/tests/test_dot.rs
expression: dot
---
digraph petgraph {
    N0[label="A"];
    N1[label="B"];
    N0 -> N1[label="EDGE_LABEL"][arrowhead="normal" dir="both" arrowtail="none"];
}

